docs/coding-guidelines.md

```diff
@@ -0,0 +1,287 @@
+# Coding Guidelines
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
+**Table of Contents**
+
+- [Coding Guidelines](#coding-guidelines)
+    - [Hierarchical State Machines](#hierarchical-state-machines)
+        - [Conventions for `poll` implementations](#conventions-for-poll-implementations)
+        - [Prioritize local work over new work from a remote](#prioritize-local-work-over-new-work-from-a-remote)
+    - [Bound everything](#bound-everything)
+        - [Channels](#channels)
+        - [Local queues](#local-queues)
+        - [Further reading](#further-reading)
+    - [No premature optimizations](#no-premature-optimizations)
+    - [Keep things sequential unless proven to be slow](#keep-things-sequential-unless-proven-to-be-slow)
+    - [Use `async/await` for sequential execution only](#use-asyncawait-for-sequential-execution-only)
+    - [Don't communicate by sharing memory; share memory by communicating.](#dont-communicate-by-sharing-memory-share-memory-by-communicating)
+        - [Further Reading](#further-reading)
+    - [Use iteration not recursion](#use-iteration-not-recursion)
+        - [Further Reading](#further-reading-1)
+
+<!-- markdown-toc end -->
+
+
+Below is a set of coding guidelines followed across the rust-libp2p code base.
+
+## Hierarchical State Machines
+
+If you sqint, rust-libp2p is just a big hierarchy of [state
+machines](https://en.wikipedia.org/wiki/Finite-state_machine) where parents pass
+events down to their children and children pass events up to their parents.
+
+![Architecture](architecture.svg)
```

Comment:
Instead of embedding the file, you could reference it planttext.com: https://www.planttext.com/api/plantuml/svg/SoWkIImgAStDuL80Wk3onA9S1PiQNLs5eFpy4gVKZCIopFpI8dH9v_oylDJaaipyl83yV0oj8KM9USK5-KKbO0aKWo0A0ZBpqb7DngB8CoKrhoGphPAWGk_4bDJSdCn4RAwrKYZ8pydHqCIY5qKArKEeSd3le0gZU09L2iFfgD8O3aP8EwJcfG0Z0m00

The same token also works as an "edit" link: https://www.planttext.com/?text=SoWkIImgAStDuL80Wk3onA9S1PiQNLs5eFpy4gVKZCIopFpI8dH9v_oylDJaaipyl83yV0oj8KM9USK5-KKbO0aKWo0A0ZBpqb7DngB8CoKrhoGphPAWGk_4bDJSdCn4RAwrKYZ8pydHqCIY5qKArKEeSd3le0gZU09L2iFfgD8O3aP8EwJcfG0Z0m00

Reply:
I am not familiar with planttext. Is it a service we can rely on? What if they prune their database?

Reply:
I've been using it for years without issues.

There is no database, I am pretty sure the diagram source code is embedded in the URL in some form of encoding.

Reply:
That makes sense. I would argue though that there is benefit in being able to access the diagram and thus read the document without internet access.